pkg/controllers/flannelmigration/ipam_migrator.go

```diff
@@ -0,0 +1,375 @@
+	// flannel.alpha.coreos.com/backend-data: '{"VtepMAC":"56:1d:8d:30:79:97"}'
+	// flannel.alpha.coreos.com/backend-type: vxlan
+	// flannel.alpha.coreos.com/public-ip: 172.16.101.96
+	backendType, ok := node.Annotations[m.config.FlannelAnnotationPreifx+"/"+flannelNodeAnnotationKeyBackendType]
+	if !ok {
+		return fmt.Errorf("node %s missing annotation for Flannel backend type", node.Name)
+	}
+	if backendType != "vxlan" {
+		return fmt.Errorf("node %s got wrong Flannel backend type %s", node.Name, backendType)
+	}
+
+	backendData, ok := node.Annotations[m.config.FlannelAnnotationPreifx+"/"+flannelNodeAnnotationKeyBackendData]
+	if !ok {
+		return fmt.Errorf("node %s missing annotation for Flannel backend data", node.Name)
+	}
+
+	publicIP, ok := node.Annotations[m.config.FlannelAnnotationPreifx+"/"+flannelNodeAnnotationKeyPublicIP]
+	if !ok {
+		return fmt.Errorf("node %s missing annotation for Flannel public ip", node.Name)
+	}
+	if _, _, err := cnet.ParseCIDROrIP(publicIP); err != nil {
+		return fmt.Errorf("node %s got wrong Flannel public ip '%s'", node.Name, publicIP)
+	}
+
+	type flannelVtepMac struct {
+		VtepMAC string
+	}
+	var fvm flannelVtepMac
+	err = json.Unmarshal([]byte(backendData), &fvm)
+	if err != nil {
+		return fmt.Errorf("node %s got wrong Flannel backend data %s", node.Name, backendData)
+	}
+
+	vtepMac := fvm.VtepMAC
+	log.Infof("node %s has vxlan setup from Flannel (vtepMac: '%s', vtepIP: '%s').", node.Name, vtepMac, vtepIP.String())
+
+	// Allocate Calico IPAM blocks for node.
+	claimed, failed, err := m.calicoClient.IPAM().ClaimAffinity(m.ctx, *cidr, node.Name)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to claim IPAM blocks for node %s, claimed %d, failed %d", node.Name, len(claimed), len(failed))
+		return err
+	}
+	log.Infof("%d IPAM blocks claimed for node %s.", len(claimed), node.Name)
+
+	// Update Calico node with Flannel vtep IP/Mac/publicIP.
+	err = setupCalicoNodeVxlan(m.ctx, m.calicoClient, node.Name, *vtepIP, vtepMac, publicIP)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Setting up Calico IPAM for node %s completed successfully.", node.Name)
+	return nil
+}
+
+// MigrateNodes setup Calico IPAM for array of nodes.
+func (m ipamMigrator) MigrateNodes(nodes []*v1.Node) error {
+	log.Infof("Start IPAM migration process for %d nodes.", len(nodes))
+	for _, node := range nodes {
+		err := m.SetupCalicoIPAMForNode(node)
+		if err != nil {
+			return err
+		}
+	}
+	log.Infof("%d nodes completed IPAM migration process.", len(nodes))
+
+	return nil
+}
+
+// setupCalicoNodeVxlan assigns specified IP/Mac address as vtep IP/Mac address for Calico node.
+func setupCalicoNodeVxlan(ctx context.Context, c client.Interface, nodeName string, vtepIP cnet.IP, mac, publicIP string) error {
+	log.Infof("Updating Calico Node %s with vtep IP %s, Mac %s.", nodeName, vtepIP.String(), mac)
+
+	// Assign vtep IP.
+	// Check current status of vtep IP. It could be assigned already.
```

Comment:
When do we expect it to be assigned already? Or is this just that theoretically it could be assigned?

Reply:
If IPAM migration is done and migration controller crashes and restarts, vtep IP for nodes will be assigned already.